BSP/STM32F413/Src/BSP_Contactor.c

```diff
@@ -1,64 +1,105 @@
+	switch (contactorChoice)
+	{
+	case ARRAY_CONTACTOR:
+		BSP_PWM_Set(ARRAY_CONTACTOR_OUT, PWM_ON_TIME);
+		break;
+	case HVHIGH_CONTACTOR:
+		BSP_PWM_Set(HVHIGH_CONTACTOR_OUT, PWM_ON_TIME);
+		break;
+	case HVLOW_CONTACTOR:
+		BSP_PWM_Set(HVLOW_CONTACTOR_OUT, PWM_ON_TIME);
+		break;
+	case ALL_CONTACTORS: //c1 and cfan are the only two possible contactors, so we operate on both
+		BSP_PWM_Set(CFAN, PWM_ON_TIME);
+		BSP_PWM_Set(C1, PWM_ON_TIME);
+		break;
+	default:
+		break;
+	}
 }
 
 /**
- * @brief   Opens the Contactor switch i.e. turns off the whole electrical system.
- * @note    May be good in the future to make this return something if the contactor could not successfully open.
- * @param   None
+ * @brief   Opens the specified Contactor switch(es) i.e. turns off the whole electrical system.
+ * @param   Contactor to turn off
  * @return  None
  */
-void BSP_Contactor_Off(void) {
-    GPIO_WriteBit(GPIOB, GPIO_Pin_0, Bit_RESET);	// Set pin 0 low
+void BSP_Contactor_Off(Contactors_t contactorChoice) {
+    // set output pin to a duty cycle of 0
+	switch (contactorChoice)
+		{
+		case ARRAY_CONTACTOR:
+			BSP_PWM_Set(ARRAY_CONTACTOR_OUT, 0);
+			break;
+		case HVHIGH_CONTACTOR:
+			BSP_PWM_Set(HVHIGH_CONTACTOR_OUT, 0);
+			break;
+		case HVLOW_CONTACTOR:
+			BSP_PWM_Set(HVLOW_CONTACTOR_OUT, 0);
+			break;
+		case ALL_CONTACTORS:
+			BSP_PWM_Set(CFAN, 0);
+			BSP_PWM_Set(C1, 0);
+			break;
+		default:
+			break;
+		}
 }
 
 /**
  * @brief   Gets the state of the Contactor switch from one of its AUX pins.
- * @param   None
+ * @note	You cannot get the state of ALL_CONTACTORS. As such, if that param is passed, it will return the state of the array contactor.
+ * @param   Contactor to get state of
  * @return  0 if contactor is off/open, 1 if on/closed
  */
-bool BSP_Contactor_GetState(void) {
-    // The AUX pin we have connected to is negative logic
-    if((GPIOB->IDR & GPIO_Pin_1) >> 1) {
-		return 0;
-	} else {
-		return 1;
+bool BSP_Contactor_GetState(Contactors_t contactorChoice) {
+	bool contactorReturnValue = ((C1_PORT->IDR & GPIO_Pin_1) >> 1) ? 0 : 1; //read the one and only input pin
+
+	/* this is old implementation, but we only have one input pin now
+	bool contactorReturnValue = false;
+	if (contactorChoice == ARRAY_CONTACTOR) {
+		contactorReturnValue = ((C1_PORT->IDR & GPIO_Pin_1) >> 1) ? 0 : 1;
+	}
+	else if (contactorChoice == HVHIGH_CONTACTOR) {
+		contactorReturnValue = ((C2_PORT->IDR & GPIO_Pin_5) >> 5) ? 0 : 1;
 	}
+	else if (contactorChoice == HVLOW_CONTACTOR) {
+		contactorReturnValue = ((C3_PORT->IDR & GPIO_Pin_1) >> 1) ? 0 : 1;
+	}
+	else if (contactorChoice == ALL_CONTACTORS) {
+		// return if ANY of the contactors are off.
+		contactorReturnValue = (((C1_PORT->IDR & GPIO_Pin_1) >> 1) ? 0 : 1)
+			&& (((C2_PORT->IDR & GPIO_Pin_5) >> 5) ? 0 : 1)
+			&& (((C3_PORT->IDR & GPIO_Pin_1) >> 1) ? 0 : 1);
+	}
+
+	*/
+	return contactorReturnValue;
```

Comment:
can also just return in one line here

Reply:
We can, but we wanna leave the old code in case we use multiple input sensing lines. Just in case. Either way would be fine imo.